// rtsp-streaming-server/build/lib/ClientWrapper.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ClientWrapper = void 0;

var _uuid = require("uuid");

var _Client = require("./Client");

var _utils = require("./utils");

const debug = (0, _utils.getDebugger)('ClientWrapper');

class ClientWrapper {
  constructor(clientServer, req) {
    this.id = (0, _uuid.v4)();
    this.clientServer = clientServer;
    this.clients = {};
    debug('%s - constructed', this.id);
    const info = (0, _utils.getMountInfo)(req.uri);
    const mount = clientServer.mounts.mounts[info.path];

    if (!mount) {
      throw new Error('Mount does not exist');
    }

    this.context = req.context || {};
    this.mount = mount;
    this.authorizationHeader = req.headers.authorization || '';
  }
  /**
   *
   * @param mounts
   * @param req
   */


  addClient(req) {
    const client = new _Client.Client(this.mount, req); // Some clients for whatever reason don't send RTSP keepalive requests
    // (Live555 streaming media as an example)
    // RTP spec says compliant clients should be sending rtcp Receive Reports (RR) to show their "liveliness"
    // So we support this as a keepalive too.

    client.rtcpServer.on('message', _buf => {
      this.keepalive();
    });
    this.clients[client.id] = client;
    debug('%s new client %s', this.id, client.id);
    return client;
  }
  /**
   *
   */


  play() {
    for (let client in this.clients) {
      this.clients[client].play();
    }

    this.keepalive();
  }
  /**
   *
   */


  close() {
    if (this.keepaliveTimeout) {
      clearTimeout(this.keepaliveTimeout);
    }

    for (let client in this.clients) {
      this.clients[client].close();
    }

    this.clientServer.clientGone(this.id);
  }
  /**
   *
   */


  keepalive() {
    if (this.keepaliveTimeout) {
      clearTimeout(this.keepaliveTimeout);
    }

    this.keepaliveTimeout = setTimeout(async () => {
      debug('%s client timeout, closing connection', this.id);

      try {
        await this.close();
      } catch (e) {// Ignore
      }
    }, 6e4); // 60 seconds (double the normal keepalive interval)
  }

}

exports.ClientWrapper = ClientWrapper;
