// rtsp-streaming-server/build/lib/Mounts.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Mounts = void 0;

var _Mount = require("./Mount");

var _utils = require("./utils");

const debug = (0, _utils.getDebugger)('Mounts');

class Mounts {
  constructor(config) {
    this.mounts = {};
    this.rtpPorts = []; // It is assumed that each start port has a correlating end port of start+1

    for (let i = config.rtpPortStart; i < config.rtpPortStart + config.rtpPortCount; i = i + 2) {
      this.rtpPorts.push(i);
    }
  }

  getMount(uri) {
    let info = (0, _utils.getMountInfo)(uri);
    return this.mounts[info.path];
  }

  addMount(uri, sdp, hooks) {
    debug('Adding mount with path %s and SDP %O', uri, sdp);
    const info = (0, _utils.getMountInfo)(uri);
    const mount = new _Mount.Mount(this, info.path, sdp, hooks);
    this.mounts[info.path] = mount;
    return mount;
  }

  getNextRtpPort() {
    debug('%d rtp ports remaining', this.rtpPorts.length - 1);
    return this.rtpPorts.shift();
  }

  returnRtpPortToPool(port) {
    debug('%d rtp ports remaining', this.rtpPorts.length + 1);
    this.rtpPorts.push(port);
  }

  deleteMount(uri) {
    debug('Removing mount with path %s', uri);
    let info = (0, _utils.getMountInfo)(uri);
    const mount = this.mounts[info.path];

    if (mount) {
      const portsFreed = mount.close();
      this.rtpPorts = this.rtpPorts.concat(portsFreed);
      this.mounts[info.path] = undefined;
      delete this.mounts[info.path];
      return true;
    }

    return false;
  }

}

exports.Mounts = Mounts;
