// rtsp-streaming-server/build/lib/RtpUdp.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RtpUdp = void 0;

var _dgram = require("dgram");

var _utils = require("./utils");

const debug = (0, _utils.getDebugger)('RtpUdp');

class RtpUdp {
  constructor(port, stream) {
    this.port = port;
    this.stream = stream;
    this.type = port % 2 ? 'rtcp' : 'rtp';
    this.server = (0, _dgram.createSocket)('udp4');
    this.server.on('message', buf => {
      for (let id in this.stream.clients) {
        let client = this.stream.clients[id]; // Differenciate rtp and rtcp so that the client object knows which port to send to

        if (this.type === 'rtcp') {
          client.sendRtcp(buf);
        } else {
          client.sendRtp(buf);
        }
      }
    });
  }

  async listen() {
    return new Promise((resolve, reject) => {
      function onError(err) {
        return reject(err);
      }

      this.server.on('error', onError);
      this.server.bind(this.port, () => {
        debug('Opened %s listener for stream %s on path %s', this.type.toUpperCase(), this.stream.id, this.stream.mount.path);
        this.server.removeListener('error', onError);
        return resolve();
      });
    });
  }

  async close() {
    return new Promise((resolve, reject) => {
      debug('Closing UDP listeners for stream %s', this.stream.id);
      this.server.close(() => {
        return resolve();
      });
    });
  }

}

exports.RtpUdp = RtpUdp;
