// rtsp-streaming-server/build/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Client", {
  enumerable: true,
  get: function () {
    return _Client.Client;
  }
});
Object.defineProperty(exports, "ClientServer", {
  enumerable: true,
  get: function () {
    return _ClientServer.ClientServer;
  }
});
Object.defineProperty(exports, "ClientServerHooksConfig", {
  enumerable: true,
  get: function () {
    return _ClientServer.ClientServerHooksConfig;
  }
});
Object.defineProperty(exports, "ClientWrapper", {
  enumerable: true,
  get: function () {
    return _ClientWrapper.ClientWrapper;
  }
});
Object.defineProperty(exports, "Mount", {
  enumerable: true,
  get: function () {
    return _Mount.Mount;
  }
});
Object.defineProperty(exports, "Mounts", {
  enumerable: true,
  get: function () {
    return _Mounts.Mounts;
  }
});
Object.defineProperty(exports, "PublishServer", {
  enumerable: true,
  get: function () {
    return _PublishServer.PublishServer;
  }
});
Object.defineProperty(exports, "PublishServerHooksConfig", {
  enumerable: true,
  get: function () {
    return _PublishServer.PublishServerHooksConfig;
  }
});
Object.defineProperty(exports, "RtpUdp", {
  enumerable: true,
  get: function () {
    return _RtpUdp.RtpUdp;
  }
});
exports.RtspServer = exports.default = void 0;

var _Client = require("./lib/Client");

var _ClientServer = require("./lib/ClientServer");

var _ClientWrapper = require("./lib/ClientWrapper");

var _Mount = require("./lib/Mount");

var _Mounts = require("./lib/Mounts");

var _PublishServer = require("./lib/PublishServer");

var _RtpUdp = require("./lib/RtpUdp");

class RtspServer {
  constructor(config) {
    this.Mounts = new _Mounts.Mounts({
      rtpPortCount: config.rtpPortCount,
      rtpPortStart: config.rtpPortStart
    });
    this.PublishServer = new _PublishServer.PublishServer(config.serverPort, this.Mounts, config.publishServerHooks);
    this.ClientServer = new _ClientServer.ClientServer(config.clientPort, this.Mounts, config.clientServerHooks);
  }

  async start() {
    try {
      await this.PublishServer.start();
      await this.ClientServer.start();
    } catch (e) {
      throw e;
    }
  }

}

exports.RtspServer = exports.default = RtspServer;
